server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -365,24 +368,139 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
+            throw new MapperParsingException("cannot merge mapping sources with different roots");
+        }
+        return (mergedRawMapping != null) ? doMerge(type, reason, mergedRawMapping) : null;
+    }
+
+    private static class RawFieldMappingMerge implements XContentHelper.CustomMerge {
+        private static final XContentHelper.CustomMerge INSTANCE = new RawFieldMappingMerge();
+
+        private static final Set<String> MERGEABLE_OBJECT_TYPES = Set.of(ObjectMapper.CONTENT_TYPE, NestedObjectMapper.CONTENT_TYPE);
+
+        private RawFieldMappingMerge() {}
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object merge(String parent, String key, Object oldValue, Object newValue) {
+            if (oldValue instanceof Map && newValue instanceof Map) {
+                if ("properties".equals(parent)) {
+                    // merging two mappings of the same field, where "key" is the field name
+                    Map<String, Object> baseMap = (Map<String, Object>) oldValue;
+                    Map<String, Object> mapToMerge = (Map<String, Object>) newValue;
+                    if (shouldMergeFieldMappings(baseMap, mapToMerge)) {
+                        // if two field mappings are to be merged, we only want to keep some specific entries from the base mapping and
+                        // let all others be overridden by the second mapping
+                        Map<String, Object> mergedMappings = new HashMap<>();
+                        if (baseMap.containsKey("properties")) {
+                            mergedMappings.put("properties", new HashMap<>((Map<String, Object>) baseMap.get("properties")));
+                        }
+                        if (baseMap.containsKey("subobjects")) {
+                            mergedMappings.put("subobjects", baseMap.get("subobjects"));
+                        }
+                        // recursively merge these two field mappings
+                        XContentHelper.merge(key, mergedMappings, mapToMerge, INSTANCE);
+                        return mergedMappings;
+                    } else {
+                        // non-mergeable types - replace the entire mapping subtree for this field
+                        return mapToMerge;
+                    }
+                }
+                // anything else (e.g. "_doc", "_meta", "properties") - no custom merge, rely on caller merge logic
+                // field mapping entries of Map type (like "fields" and "meta") are handled above and should never reach here
+                return null;
+            } else {
+                switch (key) {
+                    case "subobjects" -> {
+                        if (oldValue.equals(newValue) == false) {
+                            throw new MapperParsingException("contradicting subobjects settings provided for field: " + parent);
+                        }
+                    }
```

Comment:
I'd like to understand more why we're throwing an exception in this case. It seems to prevent the following use case:

- we set `subobjects: false` in `logs@mappings`
- user wants to use subobjects (for example to use the `nested` field type) so they set `subobjects: true` in `logs@custom`

Reply:
Since this merge is done on the raw mappings, it is only about _**explicit**_ contradiction in two field mappings, so I assumed this should lead to an error (if you set `subobjects: false` for a specific field, it is likely for a reason, so setting the opposite for the same field sounds like an error). However, I assume you are thinking about setting at the root level, where it is more likely to be a non-error merge.
Whatever all you wise reviewers here think is fine with me, I won't insist about this 🙂

Reply:
Yes, I'm mostly concerned about the case where there are different values for `subobjects` at the root level. I don't have a strong opinion on other levels. However, I'd prefer the logic to be consistent, if possible. I think I'd prefer not having that restriction on the merge algorithm itself. If the mapping is invalid due to overriding `subobjects`, it'll fail in a different place, with a more specific error message. If it's not invalid, I don't see why we'd want to disallow overriding that flag.

Reply:
I think this will be easiest to reason about if we separate out object mappers and leaf mappers:
- for leaf mappers, later templates win by replacing the definition entirely, so there's no merging
- for object mappers that have children appearing in multiple templates, the above rule means that the later templates win and replace the child definitions entirely
- given that, I think it makes most sense (and is most understandable) to also apply this to object mapper parameters, that is, later templates with parameter values that conflict with earlier templates win.  And as Felix says, if we end up with an invalid mapping at the end, we will get a failure in any case.

Reply:
Thanks for the input. I removed the explicit `subobjects` contradiction verification.